src/components/heatmap/utils.js

```diff
@@ -83,34 +85,80 @@ export function getCellByGeneTile(view, {
-  } if (viewport.id === 'axisTop') {
+  }
+
+  if (viewport.id === 'axisTop') {
     return layer.id.startsWith('axisTop');
-  } if (viewport.id === 'heatmap') {
+  }
+
+  if (viewport.id === 'cellColorLabel') {
+    return layer.id.startsWith('cellColorLabel');
+  }
+
+  if (viewport.id === 'heatmap') {
     return layer.id.startsWith('heatmap');
-  } if (viewport.id === 'colorsLeft') {
-    return layer.id.startsWith('colorsLeft');
-  } if (viewport.id === 'colorsTop') {
-    return layer.id.startsWith('colorsTop');
   }
+
+  if (viewport.id.startsWith('colorsLeft')) {
+    const matches = viewport.id.match(/-(\d)/);
+    if (matches) return layer.id.startsWith(`colorsLeftLayer-${matches[1]}`);
+  }
+
+  if (viewport.id.startsWith('colorsTop')) {
+    const matches = viewport.id.match(/-(\d)/);
+    if (matches) return layer.id.startsWith(`colorsTopLayer-${matches[1]}`);
+  }
+
   return false;
 }
 
+/**
+ * Uses canvas.measureText to compute and return the width of the given text
+ * of given font in pixels.
+ *
+ * @param {String} text The text to be rendered.
+ * @param {String} font The css font descriptor that text is to be rendered
+ * with (e.g. "bold 14px verdana").
+ *
+ * @see https://stackoverflow.com/questions/118241/calculate-text-width-with-javascript/21015393#21015393
+ */
+function getTextWidth(text, font) {
+  // re-use canvas object for better performance
+  const canvas = getTextWidth.canvas || (getTextWidth.canvas = document.createElement('canvas'));
+  const context = canvas.getContext('2d');
+  context.font = font;
+  const metrics = context.measureText(text);
+  return metrics.width;
+}
+
 /**
  * Get the size of the left and top heatmap axes,
  * taking into account the maximum label string lengths.
  * @param {boolean} transpose Is the heatmap transposed?
- * @param {number} geneLabelMaxLength What is the maximum length gene label?
- * @param {number} cellLabelMaxLength What is the maximum length cell label?
+ * @param {String} longestGeneLabel longest gene label
+ * @param {String} longestCellLabel longest cell label
+ * @param {boolean} hideObservationLabels are cell labels hidden?
+ * @param {boolean} hideVariableLabels are gene labels hidden?
+ * Increases vertical space for heatmap
  * @returns {number[]} [axisOffsetLeft, axisOffsetTop]
  */
-export function getAxisSizes(transpose, geneLabelMaxLength, cellLabelMaxLength) {
+export function getAxisSizes(
+  transpose, longestGeneLabel, longestCellLabel,
+  hideObservationLabels, hideVariableLabels,
+) {
+  const font = `${AXIS_LABEL_TEXT_SIZE}pt ${AXIS_FONT_FAMILY}`;
+  const geneLabelMaxWidth = hideVariableLabels
+    ? 0 : getTextWidth(longestGeneLabel, font) + AXIS_PADDING;
+  const cellLabelMaxWidth = hideObservationLabels
+    ? 0 : getTextWidth(longestCellLabel, font) + AXIS_PADDING;
+
   const axisOffsetLeft = clamp(
-    (transpose ? geneLabelMaxLength : cellLabelMaxLength) * AXIS_LABEL_TEXT_SIZE,
+    (transpose ? geneLabelMaxWidth : cellLabelMaxWidth),
```

Comment:
Ok I see this is where the transposition is accounted for - consider changing the variable names then be agnostic to the underlying data (i.e left/top or feature/observation).  For example the logic could be refactored so that this line is unnecessary.  I know that this component still uses a lot of the cell/gene terminology in the variable names, but we are trying to move away from that.